### PR TITLE
enhance isinf/isnan in tensor util, avoid copy back to cpu

### PR DESCRIPTION
**Pull request by dzhwinter**

### CODE REVIEW

File: paddle/fluid/framework/tensor_util.cc [hunk 1]
```diff
@@ -206,13 +207,39 @@ struct AnyVisitor : public boost::static_visitor<bool> {
   }
 };
 
+template <typename Predicate>
+struct AnyOutVisitor : public boost::static_visitor<> {
```
Comment (Superjomn) [inline paddle/fluid/framework/tensor_util.cc]: 感觉就这里可能有问题了

File: paddle/fluid/framework/tensor_util.cc [hunk 2]
```diff
@@ -206,13 +206,39 @@ struct AnyVisitor : public boost::static_visitor<bool> {
...
+  const framework::Tensor& tensor_;
+  mutable framework::Tensor* out_;
+  Predicate predicate_;
```
Comment (panyx0718) [inline paddle/fluid/framework/tensor_util.cc:216]: should these be private?

File: python/paddle/fluid/layers/tensor.py [hunk 3]
```diff
@@ -652,3 +641,33 @@ def load_combine(out, file_path):
         inputs={},
         output={"Out": out},
         args={"file_path": file_path})
+
+
+def has_inf(x):
+    """
+    Test if any of x contains a infinity number
```
Comment (panyx0718) [inline python/paddle/fluid/layers/tensor.py]: Need formal doc for public API

File: python/paddle/fluid/layers/tensor.py [hunk 4]
```diff
@@ -652,3 +641,33 @@ def load_combine(out, file_path):
...
+    """
+    helper = LayerHelper("isinf", **locals())
+    out = helper.create_tmp_variable(dtype=x.dtype)
+    helper.append_op(type="isinf", inputs={"X": x}, outputs={"Out": out})
+    return out
+
+
+def has_nan(x):
+    """
+    Test if any of x contains a NAN
```
Comment (panyx0718) [inline python/paddle/fluid/layers/tensor.py:664]: Need formal doc for public API

File: python/paddle/fluid/layers/tensor.py [hunk 5]
```diff
@@ -652,3 +641,33 @@ def load_combine(out, file_path):
...
+    helper = LayerHelper("isnan", **locals())
+    out = helper.create_tmp_variable(dtype=x.dtype)
+    helper.append_op(type="isnan", inputs={"X": x}, outputs={"Out": out})
+    return out
+
+
+def isfinite(x):
+    """
+    Test if any of x contains a infinity number
+    """
```
Comment (panyx0718) [inline python/paddle/fluid/layers/tensor.py:688]: Need formal doc for public API

File: paddle/fluid/framework/tensor_util.cc [hunk 6]
```diff
@@ -241,6 +273,71 @@ bool TensorContainsInf(const framework::Tensor& tensor) {
...
+// should be implemented if the performance hurts.
+bool TensorIsfinite(const framework::Tensor& tensor) {
+  ContainsInfPredicate pred_inf;
+  ContainsNANPredicate pred_nan;
+  return !Any(tensor, pred_inf) && !Any(tensor, pred_nan);
+}
+
+#ifdef PADDLE_WITH_CUDA
+template <typename T>
+static inline void __global__ EqualAssign(const T* cmp, T* out) {
```
Comment (panyx0718) [inline paddle/fluid/framework/tensor_util.cc]: Why is this called EqualAssign? Should it be something like "IsBothFalse"?

File: paddle/fluid/framework/tensor_util.h [hunk 7]
```diff
@@ -57,8 +57,14 @@ void TensorToVector(const Tensor& src, const platform::DeviceContext& ctx,
 template <typename T>
 void TesnorToVector(const Tensor& src, std::vector<T>* dst);
 
+// copy the result bool to cpu
 bool TensorContainsNAN(const framework::Tensor& tensor);
 bool TensorContainsInf(const framework::Tensor& tensor);
+bool TensorIsfinite(const framework::Tensor& tensor);
+
+void TensorContainsNAN(const framework::Tensor& tensor, framework::Tensor* out);
```
Comment (panyx0718) [inline paddle/fluid/framework/tensor_util.h:66]: Doc the different from previous functions?

File: paddle/fluid/framework/tensor_util.cc (hunk 2 above)
Comment (dzhwinter) [reply paddle/fluid/framework/tensor_util.cc:216]: done.

File: paddle/fluid/framework/tensor_util.h (hunk 7 above)
Comment (dzhwinter) [reply paddle/fluid/framework/tensor_util.h:66]: commented.

File: python/paddle/fluid/layers/tensor.py (hunk 3 above)
Comment (dzhwinter) [reply python/paddle/fluid/layers/tensor.py]: done.

File: python/paddle/fluid/layers/tensor.py (hunk 4 above)
Comment (dzhwinter) [reply python/paddle/fluid/layers/tensor.py:664]: done.

File: python/paddle/fluid/layers/tensor.py (hunk 5 above)
Comment (dzhwinter) [reply python/paddle/fluid/layers/tensor.py:688]: done.

File: paddle/fluid/framework/tensor_util.cc (hunk 1 above)
Comment (dzhwinter) [reply paddle/fluid/framework/tensor_util.cc]: double checked.

**Review by panyx0718 — APPROVED** [review]

**Review by typhoonzero — APPROVED** [review]
LG API

**Pull request merged**